sw/device/lib/base/csr.h

```diff
@@ -0,0 +1,287 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_BASE_CSR_H_
+#define OPENTITAN_SW_DEVICE_LIB_BASE_CSR_H_
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/stdasm.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * @file
+ * @brief Control and Status Register (CSR) interface.
+ *
+ * A set of macros that provide both read and modify operations on CSRs.
+ * Compiling translation units that include this header file with `-DMOCK_CSR`
+ * will result in the CSR operations being replaced with a mocked
+ * implementation.
+ */
+
+/**
+ * Define the implementation macros.
+ *
+ * The implementation used depends on whether the CSR library is providing a
+ * real or a mocked interface.
+ */
+#ifdef MOCK_CSR
+
+uint32_t mock_csr_read(uint32_t addr);
+
+#define CSR_READ_IMPL(csr, dst)  \
+  do {                           \
+    asm volatile("" ::"i"(csr)); \
+    *dst = mock_csr_read(csr);   \
+  } while (0);
+
+void mock_csr_write(uint32_t addr, uint32_t value);
+
+#define CSR_WRITE_IMPL(csr, val) \
+  do {                           \
+    asm volatile("" ::"i"(csr)); \
+    mock_csr_write(csr, val);    \
+  } while (0);
+
+void mock_csr_set_bits(uint32_t addr, uint32_t mask);
+
+#define CSR_SET_BITS_IMPL(csr, mask) \
+  do {                               \
+    asm volatile("" ::"i"(csr));     \
+    mock_csr_set_bits(csr, mask);    \
+  } while (0);
+
+void mock_csr_clear_bits(uint32_t addr, uint32_t mask);
+
+#define CSR_CLEAR_BITS_IMPL(csr, mask) \
+  do {                                 \
+    asm volatile("" ::"i"(csr));       \
+    mock_csr_clear_bits(csr, mask);    \
+  } while (0);
+
+#else  // MOCK_CSR
+
+#define CSR_READ_IMPL(csr, dst)                           \
+  do {                                                    \
+    asm volatile("csrr %0, %1;" : "=r"(*dst) : "i"(csr)); \
+  } while (0)
+
+#define CSR_WRITE_IMPL(csr, val)                       \
+  do {                                                 \
+    asm volatile("csrw %0, %1;" ::"i"(csr), "r"(val)); \
+  } while (0)
+
+#define CSR_SET_BITS_IMPL(csr, mask)                    \
+  do {                                                  \
+    asm volatile("csrs %0, %1;" ::"i"(csr), "r"(mask)); \
+  } while (0)
+
+#define CSR_CLEAR_BITS_IMPL(csr, mask)                  \
+  do {                                                  \
+    asm volatile("csrc %0, %1;" ::"i"(csr), "r"(mask)); \
+  } while (0)
+
+#endif  // MOCK_CSR
+
+/**
+ * CSR definitions.
```

Comment:
Is it worth documenting that these are the CSRs supported by Ibex?

Reply:
++

Reply:
I've liberally added Ibex throughout the comments. I think it's a bit clearer now.